manpages/lsdosattr.1
.TH lsdosattr 1
.SH NAME
lsdosattr \- list DOS file attributes
.SH SYNOPSIS
.B lsdosattr
[
.B \-RadlVh
]
[
.I FILES...
]
.SH DESCRIPTION
.P
.B lsdosattr
lists the DOS attributes, size and VDL(valid data length) of files. The DOS
attributes are
.IP
\fBr\fP (READONLY)
.IP
\fBh\fP (HIDDEN)
.IP
\fBs\fP (SYSTEM)
.IP
\fBv\fP (VOLUME)
.IP
\fBs\fP (DIRECTORY)
.IP
\fBa\fP (ARCHIVED)
.P
Each file is shown in 4 columns that represent the attributes, size, VDL of the
file. When the option \fB\-l\fP is used, the columns are shown in the order of file
name, size, VDL and long attributes.

.SH OPTIONS
.TP
.B \-R
List subdirectories recursively
.TP
.B \-a
Do not ignore entries starting with a dot('.')
.TP
.B \-d
List directories themselves, not their contents
.TP
.B \-l
Use long attribute names instead of single character abbreviations
.TP
.B \-V
Show version
.TP
.B \-h
Show help

.SH EXIT STATUS
.TP
.B 0
on success
.TP
.B 1
when no file could be listed
.TP
.B 2
on usage error
.TP
.B 3
when an error occurred but at least one file was successfully listed

.SH NOTE
.P
Linux kernel does not provide a special interface for reading the VDL(valid data
length). If the in-kernel exFAT implementation is "VDL-aware", the VDL offset is
treated as the start of the "hole"(\fBSEEK_HOLE\fP) that ranges to EOF.
Therefore, the approximate VDL can be inferred from \fBlseek\fP(2) and this is
why \fBlsdosattr\fP merely prints the first hole it encounters as the VDL.

.P
In vFAT, exFAT and NTFS, only the READONLY attribute is controlled through owner
write permission bit. Linux kernel always sets ARCHIVE on files it creates and
modifies. HIDDEN and SYSTEM can only be controlled through \fB ioctl_fat\fP(2).
\fBlsdosattr\fP(1) and \fBchdosattr\fP(1) are the wrappers for the ioctl.

.SH SEE ALSO
.BR lsattr (1),
.BR ioctl_fat (2)
and the documentation on
.B ATTRIB
command in Microsoft Windows and various DOS.


NEWS
exfatprogs 1.3.1 - released 2025-12-15
=====================================

NEW FEATURES :
 * fsck.exfat: support repairing the allocation bitmap size.

CHANGES :
 * exfatprogs: temporarily disable building defrag.exfat due to reported
   data loss.

BUG FIXES :
 * libexfat: fix a NULL pointer dereference in read_file_dentry_set().

exfatprogs 1.3.0 - released 2025-10-15
======================================

NEW FEATURES :
 * defrag.exfat: add a tool to defragment an exFAT
   filesystem or assess its fragmentation status.
   See a man page.

CHANGES :
 * mkfs.exfat: minimize zero-out initialization work
   in quick format mode to reduce I/O time.
 * fsck.exfat: set the entry after an unused entry as unused.

BUG FIXES :
 * mkfs.exfat: fix incorrect FAT byte length calculation.
 * mkfs.exfat: avoid setting physical sector size into sector_size field.
 * fsck.exfat: fix a memory leaks in an error path.

exfatprogs 1.2.9 - released 2025-05-12
======================================

NEW FEATURES :
 * dump.exfat: support dumping directory entry sets,
   which prints all fields of directory entries and
   cluster chains. See a man page.

CHANGES :
 * exfatprogs: update the Github action for build test
   with Debain + clang + lld.

exfatprogs 1.2.8 - released 2025-03-04
======================================

BUG FIXES :
 * dump.exfat: fix an incorrect output of an entry
   position in 32-bit system.
 * mkfs.exfat: fill an oem sector with zero instead
   of one.
 * exfatprogs: fix compilation on musl based systems
   due to loff_t type. And update the Github action
   to validate builds on the system.

exfatprogs 1.2.7 - released 2025-02-03
======================================

NEW FEATURES :
 * fsck.exfat: support repairing the upcase table.

CHANGES :
 * exfatprogs: make sure to load the tbl preprocessor
   for man pages.

BUG FIXES :
 * exfatprogs: fix a double free memory error.
 * dump.exfat: fix a constraint that volume label, bitmap,
   upcase table must be located at the beginning of a root
   directory.

exfatprogs 1.2.6 - released 2024-11-20
======================================

CHANGES :
 * exfatprogs: replace obsolete autoconf and libtool
   macros.
 * mkfs.exfat: prefer the physical block size over
   the logical block size for the exFAT sector size.
 * mkfs.exfat: add notes about the format of the volume
   GUID to the man page.
 * mkfs.exfat: fix an incorrect calculation of the number
   of used clusters.

BUG FIXES :
 * exfatlabel: fix an user input error when setting
   a volume serial or label.

exfatprogs 1.2.5 - released 2024-08-06
======================================

CHANGES :
 * exfatprogs: remove the limitation that the device
   path length cannot exceed 254 bytes.
 * exfatprogs: include the test images in the release
   package.

NEW FEATURES :
 * fsck.exfat: check and repair the filename which has
   invalid characters.

BUG FIXES :
 * tune.exfat: check whether the volume has invalid
   characters correctly.
 * fsck.exfat: check whether the filename and volume
   has invalid characters correctly.
 * fsck.exfat: fix endianess issues which happen
   in the big-endian system.

exfatprogs 1.2.4 - released 2024-06-17
======================================

BUG FIXES :
 * tune.exfat: Fix "invalid serial number" error when
   setting an serial number.
 * fsck.exfat: Fix memory leak in an error path

exfatprogs 1.2.3 - released 2024-05-23
======================================

CHANGES :
 * dump.exfat: Report sector size in bytes and cluster size in
   terms of sectors.
 * fsck.exfat: Show checksum value if the SetChecksum of File
   directory entry is invalid.
 * mkfs.exfat: Improve FAT length calculation to reduce
   the FAT size.

NEW FEATURES :
 * mkfs.exfat: Add the option "--sector-size".
 * fsck.exfat: Support checking and repairing VendorAllcation and
   VendorExtension directory entries.

BUG FIXES :
 * exfatprogs: Remove unnecessary memory allocations.
 * fsck.exfat: Fix corruption that can occur if the cluster size
   is 512-byte.
 * fsck.exfat: Fix the SecondaryCount of File directory entry
   when the count of Name directory entries is 17 or higher.
 * tune.exfat: Fix an error that accepts invalid serial numbers.

exfatprogs 1.2.2 - released 2023-10-26
======================================

CHANGES :
 * exfat2img: Allow dumps for read-only devices.
 * fsck.exfat: Revert Repairing zero size directory.

NEW FEATURES :
 * fsck.exfat: Repair duplicated filename.
 * mkfs.exfat: Add the option "q" to print only error messages.
 * mkfs.exfat: Add the option "U" to set volume GUID.
 * tune.exfat: Add the option "U" / "-u" to set or print volume GUID.

BUG FIXES:
 * fsck.exfat: Fix some out-of-bounds memory accesses.
 * fsck.exfat: Change not to delete volume GUID directory entry.

exfatprogs 1.2.1 - released 2023-05-17
======================================

CHANGES :
 * fsck.exfat: Repair zero size directory.
 * fsck.exfat: Four small clean-ups.

exfatprogs 1.2.0 - released 2022-10-28
======================================

CHANGES :
 * fsck.exfat: Keep traveling files even if there is a corrupted
directory entry set.
 * fsck.exfat: Introduce the option "b" to recover a boot sector even
if an exFAT filesystem is not found.
 * fsck.exfat: Introduce the option "s" to create files in
"/LOST+FOUND", which have clusters allocated but was not belonged to
any files.
 * fsck.exfat: Rename '.' and '..' entry name to the one user want.

NEW FEATURES :
 * fsck.exfat: Repair corruptions of an exFAT filesystem. Please refer
to fsck.exfat manpage to see what kind of corruptions can be repaired.
 * exfat2img: Dump metadata of an exFAT filesystem. Please refer to
exfat2img manpage to see how to use it.

BUG FIXES:
 * fsck.exfat: Fix an infinite loop while traveling files.
 * tune.exfat: Fix bitmap entry corruption when adding new volume lablel.

exfatprogs 1.1.3 - released 2021-11-11
======================================

CHANGES :

dosattr/lsdosattr.c
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 *   Copyright (C) 2026 David Timber <dxdt@dev.snart.me>
 */

#include <stdbool.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <locale.h>
#include <getopt.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <unistd.h>

#include "dosattr.h"

const char *program_name = "lsdosattr";

struct attrm {
	const char *ro; /* ATTR_READONLY */
	const char *hi; /* ATTR_HIDDEN   */
	const char *sy; /* ATTR_SYSTEM   */
	const char *vo; /* ATTR_VOLUME   */
	const char *su; /* ATTR_SUBDIR   */
	const char *ar; /* ATTR_ARCHIVE  */
};

static struct {
	bool recursive;
	bool version_only;
	bool all;
	bool dirs;
	bool long_attrs;
	bool verbose;
	bool help;
} ui;

static const struct attrm attrm_short = {
	.ro = "r",
	.hi = "h",
	.sy = "s",
	.vo = "v",
	.su = "d",
	.ar = "a",
};
static const struct attrm attrm_long = {
	.ro = "READONLY ",
	.hi = "HIDDEN ",
	.sy = "SYSTEM ",
	.vo = "VOLUME ",
	.su = "DIRECTORY ",
	.ar = "ARCHIVED ",
};
static const struct attrm *attrm;

static void usage(void)
{
	fprintf(stderr, "Usage: %s [OPTION] FILE...\n", program_name);
	fprintf(stderr, "List the DOS attributes of each FILE\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "\t-R  list subdirectories recursively\n");
	fprintf(stderr, "\t-a  do not ignore entries starting with a dot(.)\n");
	fprintf(stderr, "\t-d  list directories themselves, not their contents\n");
	fprintf(stderr,
		"\t-l  use long attribute names instead of single character abbreviations\n");
	fprintf(stderr, "\t-V  show version\n");
	fprintf(stderr, "\t-h  show help\n");

	exit(DOSATTR_EXIT_USAGE);
}

static void push_attrs(char *out, size_t *p, const char *a)
{
	const size_t al = strlen(a);

	memcpy(out + *p, a, al);
	*p += al;
}

static int list_attributes(FTSENT *ent, bool *skip)
{
	char attrs[sizeof("read-only hidden system volume directory archived ")];
	char vdls[21];
	size_t ap = 0;
	uint32_t dosattrs = UINT32_MAX;
	int fd;
	int err = 0;
#define PUSH_ATTRS(x) push_attrs(attrs, &ap, x)

	if (!ui.recursive) {
		if (ui.dirs || (S_ISDIR(ent->fts_statp->st_mode) && ent->fts_level > 0)) {
			/* Don't visit the children, but the arg itself shouldn't be skipped */
			*skip = true;
		}
	}
	if (!ui.all && ent->fts_level > 0 && ent->fts_name[0] == '.') {
		/* Skip the "hidden" node */
		if (strcmp(ent->fts_name, ".") != 0 && strcmp(ent->fts_name, "..") != 0)
			/* Don't touch SEEDOTS, though */
			*skip = true;
		return 0;
	}

	/*
	 * FIXME: should do fstatfs() to get fstype before doing ioctl()? Should
	 * I also work on NTFS in the future?
	 */

	fd = open(ent->fts_accpath, O_RDONLY);
	if (fd < 0) {
		exfat_err("%s: %s: %s\n", program_name, ent->fts_path, strerror(errno));
		return 1;
	}

	if (ui.long_attrs)
		attrs[0] = 0;
	else
		strncpy(attrs, "******", sizeof(attrs));

	if (ioctl(fd, FAT_IOCTL_GET_ATTRIBUTES, &dosattrs) == 0) {
		if (dosattrs & ATTR_READONLY)
			PUSH_ATTRS(attrm->ro);
		else if (!ui.long_attrs)
			PUSH_ATTRS("-");

		if (dosattrs & ATTR_HIDDEN)
			PUSH_ATTRS(attrm->hi);
		else if (!ui.long_attrs)
			PUSH_ATTRS("-");

		if (dosattrs & ATTR_SYSTEM)
			PUSH_ATTRS(attrm->sy);
		else if (!ui.long_attrs)
			PUSH_ATTRS("-");

		if (dosattrs & ATTR_VOLUME)
			PUSH_ATTRS(attrm->vo);
		else if (!ui.long_attrs)
			PUSH_ATTRS("-");

		if (dosattrs & ATTR_SUBDIR)
			PUSH_ATTRS(attrm->su);
		else if (!ui.long_attrs)
			PUSH_ATTRS("-");

		if (dosattrs & ATTR_ARCHIVE)
			PUSH_ATTRS(attrm->ar);
		else if (!ui.long_attrs)
			PUSH_ATTRS("-");

		attrs[ap] = 0;
	} else if (errno != ENOTTY) {
		err = 2;
		exfat_err("%s: FAT_IOCTL_GET_ATTRIBUTES, %s: %s\n",
			program_name, ent->fts_path, strerror(errno));
	}

	snprintf(vdls, sizeof(vdls), "%c", '*');
	/*
	 * The Linux community has decided that instead of introducing a new
	 * IOCTL, the kernel should provide the VDL(valid data length)
	 * information via SEEK_DATA/SEEK_HOLE interface. If the in-kernel exFAT
	 * lseek() interface is "VDL-aware"(upcoming iomap patches), when the
	 * VDL falls short of st->st_size, the VDL can be inferred from the fact
	 * that SEEK_HOLE offset is less than EOF.
	 *
	 * Link: https://lore.kernel.org/linux-fsdevel/20260311222613.2010177-1-dxdt@dev.snart.me/
	 *
	 * Note that the VDL is in bytes but the kernel could round it up to
	 * prevent inadvertent misaligned I/O.
	 */
	if (S_ISREG(ent->fts_statp->st_mode)) {
		off_t vdl = lseek(fd, 0, SEEK_HOLE);

		if (vdl < 0 && errno == ENXIO)
			vdl = 0;

		if (vdl >= 0) {
			if (!ui.verbose && vdl == ent->fts_statp->st_size)
				snprintf(vdls, sizeof(vdls), "%c", '=');
			else
				snprintf(vdls, sizeof(vdls), "%"PRIu64, (uint64_t)vdl);
		} else if (errno != ENXIO) {
			err = 2;
			exfat_err("%s: lseek(), %s: %s\n", program_name,
				  ent->fts_path, strerror(errno));
		}
	}

	if (ui.long_attrs)
		printf("%-28s %12"PRIu64" %12s %s\n",
			ent->fts_path, ent->fts_statp->st_size, vdls, attrs);
	else
		printf("%s %12"PRIu64" %12s %s\n",
			attrs, ent->fts_statp->st_size, vdls, ent->fts_path);

	close(fd);
	return err;
#undef PUSH_ATTR
}

int main(int argc, char *argv[])
{
#define MY_FTSOPTS (FTS_PHYSICAL | FTS_SEEDOT)
	static int c, retval;

	setlocale(LC_MESSAGES, "");
	setlocale(LC_CTYPE, "");

	if (!(argc && *argv))
		usage();

	while ((c = getopt(argc, argv, "RVvadlh")) != EOF) {
		switch (c) {
		case 'R':
			ui.recursive = true;
			break;
		case 'V':
			ui.version_only = true;
			break;
		case 'v':
			ui.verbose = true;
			break;
		case 'a':
			ui.all = true;
			break;
		case 'd':
			ui.dirs = true;
			break;
		case 'l':
			ui.long_attrs = true;
			break;
		case 'h':
			ui.help = true;
			break;
		default:
			usage();
		}
	}

	if (ui.version_only)
		show_version();
	if (ui.help)
		usage();
	if (ui.help || ui.version_only)
		exit(DOSATTR_EXIT_USAGE);

	if (ui.long_attrs)
		attrm = &attrm_long;
	else
		attrm = &attrm_short;

	if (optind > argc - 1) {
		static const char *const DEFAULT_ARGV[] = { ".", NULL };

		retval = dosattr_fts((char * const *)DEFAULT_ARGV,
				list_attributes, MY_FTSOPTS);
	} else
		retval = dosattr_fts((char * const *)(argv + optind),
				list_attributes, MY_FTSOPTS);

	return retval;
}
